# 빅데이터 분석기사 실기 제 1유형
### mtcars 데이터셋을 활용한 추가 문제
DataUrl = https://raw.githubusercontent.com/Datamanim/dataq/main/mtcars.csv

In [1]:
import pandas as pd
import sklearn
import scipy

In [2]:
DataUrl = 'https://raw.githubusercontent.com/Datamanim/dataq/main/mtcars.csv'
df = pd.read_csv(DataUrl)

In [3]:
df

Unnamed: 0   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  \
0             Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1   
1         Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1   
2            Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1   
3        Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0   
4     Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0   
5               Valiant  18.1    6  225.0  105  2.76  3.460  20.22   1   0   
6            Duster 360  14.3    8  360.0  245  3.21  3.570  15.84   0   0   
7             Merc 240D  24.4    4  146.7   62  3.69  3.190  20.00   1   0   
8              Merc 230  22.8    4  140.8   95  3.92  3.150  22.90   1   0   
9              Merc 280  19.2    6  167.6  123  3.92  3.440  18.30   1   0   
10            Merc 280C  17.8    6  167.6  123  3.92  3.440  18.90   1   0   
11           Merc 450SE  16.4    8  275.8  180  3.07  4.070  17.40   0   0   
12           Merc 450SL  17.3    8  275.8  180  3.07  3.730  17.60   0   0   
13          Merc 450SLC  15.2    8  275.8  180  3.07  3.780  18.00   0   0   
14   Cadillac Fleetwood  10.4    8  472.0  205  2.93  5.250  17.98   0   0   
15  Lincoln Continental  10.4    8  460.0  215  3.00  5.424  17.82   0   0   
16    Chrysler Imperial  14.7    8  440.0  230  3.23  5.345  17.42   0   0   
17             Fiat 128  32.4    4   78.7   66  4.08  2.200  19.47   1   1   
18          Honda Civic  30.4    4   75.7   52  4.93  1.615  18.52   1   1   
19       Toyota Corolla  33.9    4   71.1   65  4.22  1.835  19.90   1   1   
20        Toyota Corona  21.5    4  120.1   97  3.70  2.465  20.01   1   0   
21     Dodge Challenger  15.5    8  318.0  150  2.76  3.520  16.87   0   0   
22          AMC Javelin  15.2    8  304.0  150  3.15  3.435  17.30   0   0   
23           Camaro Z28  13.3    8  350.0  245  3.73  3.840  15.41   0   0   
24     Pontiac Firebird  19.2    8  400.0  175  3.08  3.845  17.05   0   0   
25            Fiat X1-9  27.3    4   79.0   66  4.08  1.935  18.90   1   1   
26        Porsche 914-2  26.0    4  120.3   91  4.43  2.140  16.70   0   1   
27         Lotus Europa  30.4    4   95.1  113  3.77  1.513  16.90   1   1   
28       Ford Pantera L  15.8    8  351.0  264  4.22  3.170  14.50   0   1   
29         Ferrari Dino  19.7    6  145.0  175  3.62  2.770  15.50   0   1   
30        Maserati Bora  15.0    8  301.0  335  3.54  3.570  14.60   0   1   
31           Volvo 142E  21.4    4  121.0  109  4.11  2.780  18.60   1   1   

    gear  carb  
0      4     4  
1      4     4  
2      4     1  
3      3     1  
4      3     2  
5      3     1  
6      3     4  
7      4     2  
8      4     2  
9      4     4  
10     4     4  
11     3     3  
12     3     3  
13     3     3  
14     3     4  
15     3     4  
16     3     4  
17     4     1  
18     4     2  
19     4     1  
20     3     1  
21     3     2  
22     3     2  
23     3     4  
24     3     2  
25     4     1  
26     5     2  
27     5     2  
28     5     4  
29     5     6  
30     5     8  
31     4     2

In [4]:
df.dtypes

Unnamed: 0     object
mpg           float64
cyl             int64
disp          float64
hp              int64
drat          float64
wt            float64
qsec          float64
vs              int64
am              int64
gear            int64
carb            int64
dtype: object

### 문제1) 
연비 mpg 변수 관련하여 단일표본 T검정 수행하고자 한다. 먼저 데이터가 정규성을

충족하는지 확인하기 위해 sharpiro 테스트를 진행 하고 이를 통해 나오는  p-value 값을 소수 

셋째자리까지 구하시오(넷째자리에서 버림)
[출처] [빅데이터 분석기사 실기] 통계 문제 5선|작성자 마스터 킴

In [5]:
help(scipy.stats.shapiro)

Help on function shapiro in module scipy.stats.morestats:

shapiro(x)
    Perform the Shapiro-Wilk test for normality.
    
    The Shapiro-Wilk test tests the null hypothesis that the
    data was drawn from a normal distribution.
    
    Parameters
    ----------
    x : array_like
        Array of sample data.
    
    Returns
    -------
    statistic : float
        The test statistic.
    p-value : float
        The p-value for the hypothesis test.
    
    See Also
    --------
    anderson : The Anderson-Darling test for normality
    kstest : The Kolmogorov-Smirnov test for goodness of fit.
    
    Notes
    -----
    The algorithm used is described in [4]_ but censoring parameters as
    described are not implemented. For N > 5000 the W test statistic is accurate
    but the p-value may not be.
    
    The chance of rejecting the null hypothesis when it is true is close to 5%
    regardless of sample size.
    
    References
    ----------
    .. [1] https://www.itl.nist.

In [6]:
from scipy import stats
result = shapiro_test = stats.shapiro(df['mpg'])

In [7]:
print(result)

ShapiroResult(statistic=0.9475647211074829, pvalue=0.12288112193346024)


p-value가 0.122이다. 0.05보다 크므로 귀무가설 채택, 정규성 충족

### 문제2) 
자동차 구매시 연비를 우선순위로 삼고자 한다. Mazda Rx4를 구매하고자 하는데 
이 차량의 연비는 21.0이다. 이 연비 수준이 차량들의 평균 연비와 비슷한 수준인지 단일표본 T검정을 실시한다. 21에 대한 단일표본 p-value값을 소수 셋째자리까지 구하시오.(넷째자리에서 버림)

​
[출처] [빅데이터 분석기사 실기] 통계 문제 5선|작성자 마스터 킴

In [8]:
help(scipy.stats.ttest_1samp)

Help on function ttest_1samp in module scipy.stats.stats:

ttest_1samp(a, popmean, axis=0, nan_policy='propagate', alternative='two-sided')
    Calculate the T-test for the mean of ONE group of scores.
    
    This is a two-sided test for the null hypothesis that the expected value
    (mean) of a sample of independent observations `a` is equal to the given
    population mean, `popmean`.
    
    Parameters
    ----------
    a : array_like
        Sample observation.
    popmean : float or array_like
        Expected value in null hypothesis. If array_like, then it must have the
        same shape as `a` excluding the axis dimension.
    axis : int or None, optional
        Axis along which to compute test; default is 0. If None, compute over
        the whole array `a`.
    nan_policy : {'propagate', 'raise', 'omit'}, optional
        Defines how to handle when input contains nan.
        The following options are available (default is 'propagate'):
    
          * 'propagate': re

In [9]:
from scipy import stats
result = stats.ttest_1samp(df['mpg'],21)

In [10]:
print(result)

Ttest_1sampResult(statistic=-0.8535334614920062, pvalue=0.39991093888673557)


p-value가 약 0. 4, 즉 0.05보다 크기 떄문에 귀무가설 채택, 평균과 큰 차이가 없다.

### 문제3) 
오토 방식과 수동 방식간의 연비 차이가 통계적으로 유의미한지 확인하고자 한다.
오토방식은 am 변수의 1인 값이며 수동은 am 변수의 0이라 할 수 있다. 두집단의 평균 차이를 독립표본 T검정을 수행하기 앞서 두 집단의 아웃라이어를 IQR 방식을 활용하여 제거한 후 수행한다.(3분위 + IQR*1.5 / 1분위 - IQR*1.5) 이때의 T-value 값의 절대값을 소수 셋째자리를 버림해 둘째자리 까지 구하시오(오토 방식은 13개 차량, 수동은 19개 차량이며 등분산은 우선 충족한다고 가정한다)
[출처] [빅데이터 분석기사 실기] 통계 문제 5선|작성자 마스터 킴

In [11]:
df3_auto = df.loc[df['am']==1].sort_values('mpg').reset_index(drop=True)



In [12]:
df3_auto

Unnamed: 0   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0    Maserati Bora  15.0    8  301.0  335  3.54  3.570  14.60   0   1     5   
1   Ford Pantera L  15.8    8  351.0  264  4.22  3.170  14.50   0   1     5   
2     Ferrari Dino  19.7    6  145.0  175  3.62  2.770  15.50   0   1     5   
3        Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
4    Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
5       Volvo 142E  21.4    4  121.0  109  4.11  2.780  18.60   1   1     4   
6       Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
7    Porsche 914-2  26.0    4  120.3   91  4.43  2.140  16.70   0   1     5   
8        Fiat X1-9  27.3    4   79.0   66  4.08  1.935  18.90   1   1     4   
9      Honda Civic  30.4    4   75.7   52  4.93  1.615  18.52   1   1     4   
10    Lotus Europa  30.4    4   95.1  113  3.77  1.513  16.90   1   1     5   
11        Fiat 128  32.4    4   78.7   66  4.08  2.200  19.47   1   1     4   
12  Toyota Corolla  33.9    4   71.1   65  4.22  1.835  19.90   1   1     4   

    carb  
0      8  
1      4  
2      6  
3      4  
4      4  
5      2  
6      1  
7      2  
8      1  
9      2  
10     2  
11     1  
12     1

In [13]:
import numpy as np

auto_q3 = np.percentile(df3_auto['mpg'], 75)
auto_q1 = np.percentile(df3_auto['mpg'], 25)

In [14]:
auto_q3

30.4

In [15]:
auto_q1

21.0

In [16]:
auto_iqr = auto_q3-auto_q1

In [17]:
auto_iqr

9.399999999999999

In [18]:
auto_after = df3_auto.loc[(df3_auto['mpg'] < (auto_q3 + auto_iqr*1.5)) & (df3_auto['mpg'] > (auto_q3 - auto_iqr*1.5))]

In [19]:
auto_after

Unnamed: 0   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
2     Ferrari Dino  19.7    6  145.0  175  3.62  2.770  15.50   0   1     5   
3        Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
4    Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
5       Volvo 142E  21.4    4  121.0  109  4.11  2.780  18.60   1   1     4   
6       Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
7    Porsche 914-2  26.0    4  120.3   91  4.43  2.140  16.70   0   1     5   
8        Fiat X1-9  27.3    4   79.0   66  4.08  1.935  18.90   1   1     4   
9      Honda Civic  30.4    4   75.7   52  4.93  1.615  18.52   1   1     4   
10    Lotus Europa  30.4    4   95.1  113  3.77  1.513  16.90   1   1     5   
11        Fiat 128  32.4    4   78.7   66  4.08  2.200  19.47   1   1     4   
12  Toyota Corolla  33.9    4   71.1   65  4.22  1.835  19.90   1   1     4   

    carb  
2      6  
3      4  
4      4  
5      2  
6      1  
7      2  
8      1  
9      2  
10     2  
11     1  
12     1

In [20]:
df3_manu = df.loc[df['am'] == 0].sort_values('mpg').reset_index(drop=True)


In [21]:
df3_manu

Unnamed: 0   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  \
0    Cadillac Fleetwood  10.4    8  472.0  205  2.93  5.250  17.98   0   0   
1   Lincoln Continental  10.4    8  460.0  215  3.00  5.424  17.82   0   0   
2            Camaro Z28  13.3    8  350.0  245  3.73  3.840  15.41   0   0   
3            Duster 360  14.3    8  360.0  245  3.21  3.570  15.84   0   0   
4     Chrysler Imperial  14.7    8  440.0  230  3.23  5.345  17.42   0   0   
5           AMC Javelin  15.2    8  304.0  150  3.15  3.435  17.30   0   0   
6           Merc 450SLC  15.2    8  275.8  180  3.07  3.780  18.00   0   0   
7      Dodge Challenger  15.5    8  318.0  150  2.76  3.520  16.87   0   0   
8            Merc 450SE  16.4    8  275.8  180  3.07  4.070  17.40   0   0   
9            Merc 450SL  17.3    8  275.8  180  3.07  3.730  17.60   0   0   
10            Merc 280C  17.8    6  167.6  123  3.92  3.440  18.90   1   0   
11              Valiant  18.1    6  225.0  105  2.76  3.460  20.22   1   0   
12    Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0   
13     Pontiac Firebird  19.2    8  400.0  175  3.08  3.845  17.05   0   0   
14             Merc 280  19.2    6  167.6  123  3.92  3.440  18.30   1   0   
15       Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0   
16        Toyota Corona  21.5    4  120.1   97  3.70  2.465  20.01   1   0   
17             Merc 230  22.8    4  140.8   95  3.92  3.150  22.90   1   0   
18            Merc 240D  24.4    4  146.7   62  3.69  3.190  20.00   1   0   

    gear  carb  
0      3     4  
1      3     4  
2      3     4  
3      3     4  
4      3     4  
5      3     2  
6      3     3  
7      3     2  
8      3     3  
9      3     3  
10     4     4  
11     3     1  
12     3     2  
13     3     2  
14     4     4  
15     3     1  
16     3     1  
17     4     2  
18     4     2

In [22]:
import numpy as np

df3_manu_q3 = np.percentile(df3_manu['mpg'], 75)
df3_manu_q1 = np.percentile(df3_manu['mpg'], 25)

In [23]:
df3_manu_q3

19.2

In [24]:
df3_manu_q1

14.95

In [25]:
df3_manu_iqr = df3_manu_q3-df3_manu_q1

In [26]:
df3_manu_iqr

4.25

In [27]:
df3_manu_after = df.loc[(df3_manu['mpg'] < (df3_manu_q3 + df3_manu_iqr*1.5)) & (df['mpg'] > (df3_manu_q1 - df3_manu_iqr*1.5))]

In [28]:
df3_manu_after

Unnamed: 0   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  \
0             Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1   
1         Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1   
2            Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1   
3        Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0   
4     Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0   
5               Valiant  18.1    6  225.0  105  2.76  3.460  20.22   1   0   
6            Duster 360  14.3    8  360.0  245  3.21  3.570  15.84   0   0   
7             Merc 240D  24.4    4  146.7   62  3.69  3.190  20.00   1   0   
8              Merc 230  22.8    4  140.8   95  3.92  3.150  22.90   1   0   
9              Merc 280  19.2    6  167.6  123  3.92  3.440  18.30   1   0   
10            Merc 280C  17.8    6  167.6  123  3.92  3.440  18.90   1   0   
11           Merc 450SE  16.4    8  275.8  180  3.07  4.070  17.40   0   0   
12           Merc 450SL  17.3    8  275.8  180  3.07  3.730  17.60   0   0   
13          Merc 450SLC  15.2    8  275.8  180  3.07  3.780  18.00   0   0   
14   Cadillac Fleetwood  10.4    8  472.0  205  2.93  5.250  17.98   0   0   
15  Lincoln Continental  10.4    8  460.0  215  3.00  5.424  17.82   0   0   
16    Chrysler Imperial  14.7    8  440.0  230  3.23  5.345  17.42   0   0   
17             Fiat 128  32.4    4   78.7   66  4.08  2.200  19.47   1   1   
18          Honda Civic  30.4    4   75.7   52  4.93  1.615  18.52   1   1   

    gear  carb  
0      4     4  
1      4     4  
2      4     1  
3      3     1  
4      3     2  
5      3     1  
6      3     4  
7      4     2  
8      4     2  
9      4     4  
10     4     4  
11     3     3  
12     3     3  
13     3     3  
14     3     4  
15     3     4  
16     3     4  
17     4     1  
18     4     2

In [29]:
import scipy
import numpy as np
# help(scipy.stats.ttest_ind)
'''
 >>> rvs1 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
    >>> rvs2 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
    >>> stats.ttest_ind(rvs1, rvs2)
    Ttest_indResult(statistic=-0.4390847099199348, pvalue=0.6606952038870015)
    >>> stats.ttest_ind(rvs1, rvs2, equal_var=False)
    Ttest_indResult(statistic=-0.4390847099199348, pvalue=0.6606952553131064)
    
'''

stats.ttest_ind(auto_after['mpg'], df3_manu_after['mpg'])

Ttest_indResult(statistic=3.157401217404067, pvalue=0.003791794019621543)

3.16, p-value는 0.05보다 현저히 낮으므로, 오토와 수동의 차이에 따른 연비의 차이가 유의미 하다. 대립 채택

### 문제4) 
다음 실린더 개수별 연비차이가 통계적으로 유의미한지를 ANOVA 분석을 통해 확인한다. cyl 변수 실린더 개수를 바탕으로 3가지 집단의 연비 값에 대해 ANOVA 분석을 수행하고 F값을 소수 셋째자리에서 버림하여 둘째자리까지 나타내시오(정규성, 등분산은 우선 충족한다고 가정한다)
[출처] [빅데이터 분석기사 실기] 통계 문제 5선|작성자 마스터 킴

In [30]:
from scipy.stats import f_oneway
# help(scipy.stats.f_oneway)
'''
    Examples
    --------
    >>> from scipy.stats import f_oneway
    
    Here are some data [3]_ on a shell measurement (the length of the anterior
    adductor muscle scar, standardized by dividing by length) in the mussel
    Mytilus trossulus from five locations: Tillamook, Oregon; Newport, Oregon;
    Petersburg, Alaska; Magadan, Russia; and Tvarminne, Finland, taken from a
    much larger data set used in McDonald et al. (1991).
    
    >>> tillamook = [0.0571, 0.0813, 0.0831, 0.0976, 0.0817, 0.0859, 0.0735,
    ...              0.0659, 0.0923, 0.0836]
    >>> newport = [0.0873, 0.0662, 0.0672, 0.0819, 0.0749, 0.0649, 0.0835,
    ...            0.0725]
    >>> petersburg = [0.0974, 0.1352, 0.0817, 0.1016, 0.0968, 0.1064, 0.105]
    >>> magadan = [0.1033, 0.0915, 0.0781, 0.0685, 0.0677, 0.0697, 0.0764,
    ...            0.0689]
    >>> tvarminne = [0.0703, 0.1026, 0.0956, 0.0973, 0.1039, 0.1045]
    >>> f_oneway(tillamook, newport, petersburg, magadan, tvarminne)
    F_onewayResult(statistic=7.121019471642447, pvalue=0.0002812242314534544)
    
    `f_oneway` accepts multidimensional input arrays.  When the inputs
    are multidimensional and `axis` is not given, the test is performed
    along the first axis of the input arrays.  For the following data, the
    test is performed three times, once for each column.
    
    >>> a = np.array([[9.87, 9.03, 6.81],
    ...               [7.18, 8.35, 7.00],
    ...               [8.39, 7.58, 7.68],
    ...               [7.45, 6.33, 9.35],
    ...               [6.41, 7.10, 9.33],
    ...               [8.00, 8.24, 8.44]])
    >>> b = np.array([[6.35, 7.30, 7.16],
    ...               [6.65, 6.68, 7.63],
    ...               [5.72, 7.73, 6.72],
    ...               [7.01, 9.19, 7.41],
    ...               [7.75, 7.87, 8.30],
    ...               [6.90, 7.97, 6.97]])
    >>> c = np.array([[3.31, 8.77, 1.01],
    ...               [8.25, 3.24, 3.62],
    ...               [6.32, 8.81, 5.19],
    ...               [7.48, 8.83, 8.91],
    ...               [8.59, 6.01, 6.07],
    ...               [3.07, 9.72, 7.48]])
    >>> F, p = f_oneway(a, b, c)
    >>> F
    array([1.75676344, 0.03701228, 3.76439349])
    >>> p
    array([0.20630784, 0.96375203, 0.04733157])
    '''

df_cyl_4 = df.loc[df['cyl']==4]
df_cyl_6 = df.loc[df['cyl']==6]
df_cyl_8 = df.loc[df['cyl']==8]



In [31]:
F, p = f_oneway(df_cyl_4['mpg'], df_cyl_6['mpg'], df_cyl_8['mpg'])

In [32]:
print('F- Statistics {}, p-value {}'.format(F,p))

F- Statistics 39.697515255869035, p-value 4.978919174400203e-09


f통계량은 39.69, p-value는 4.978919174400203e-09으로 0.05보다 현저히 낮기 때문에, 대립가설을 채택한다. 즉, 실린더 개수별 연비차이가 유의미하다.

### 문제5) 
실린더의 개수가 차량의 가속도 능력과 연관성이 있는지를 카이스퀘어 검정을 통해 확인하고자 한다. 실린더 개수는 cyl변수를 활용하고 가속도 능력은 ¼ 마일에 도달하는 시간 변수인 qsec을 활용하여 qsec의 하위 25%인 16.9 이하면 하 , 상위 25%인 18.9 이상이면 상 나머지는 중으로 명목척도화 한다. 이후 실린더 개수와 차량 가속도 능력에 대한 카이스퀘어 검정을 통해 연관성을 확인한다. p-value를 소수 둘째자리까지 구하시오(셋째자리에서 버림)  
[출처] [빅데이터 분석기사 실기] 통계 문제 5선|작성자 마스터 킴

In [33]:
from scipy.stats import chi2_contingency
import numpy as np
import pandas as pd

In [34]:
df['speed'] = np.where( df['qsec'] < 16.9 , '하' , np.where(df['qsec'] <18.9, '중', '상'  ))

In [35]:
df_chai = pd.crosstab( df['speed'], df['cyl'])

In [36]:
df_chai

cyl    4  6  8
speed         
상      6  3  0
중      4  2  9
하      1  2  5

In [37]:
chi2_contingency(df_chai)

(10.664935064935065,
 0.030599318034876272,
 4,
 array([[3.09375, 1.96875, 3.9375 ],
        [5.15625, 3.28125, 6.5625 ],
        [2.75   , 1.75   , 3.5    ]]))

p-value가 0.05보다 작은 0.03이므로 대립가설을 채택한다.